# Hacker News Posts

Hacker News is a site started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result.

You can find the data set here, but note that it has been reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions.

We're specifically interested in posts whose titles begin with either Ask HN or Show HN. Users submit Ask HN posts to ask the Hacker News community a specific question. Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting.

We'll compare these two types of posts to determine the following:

1. Do Ask HN or Show HN receive more comments on average?
2. Do posts created at a certain time receive more comments on average?

In [1]:
from csv import reader
open_file = open('hacker_news.csv')
read_file = reader(open_file)
hn = list(read_file)

# Remove headers from the list
headers = hn[0]
hn = hn[1:]

print('These are the headers: ' + str(headers))
print('\n')
print('First 5 rows of hn dataset:')
print(hn[0:5])
print('\n')
print('Length of hn: '+str(len(hn)))

These are the headers: ['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


First 5 rows of hn dataset:
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '

# Extracting Ask HN and Show HN Posts

First, identify posts that begin with either Ask HN or Show HN. Separating the two types will later help us determine which type of post should we focus the analysis on.

In [10]:
# Identify posts that begin with either `Ask HN` or `Show HN`
# Then separate them into different lists
ask_posts = []
show_posts = []
other_posts = []

for post in hn:
    title = str(post[1])
    if title.lower().startswith('ask hn'):
        ask_posts.append(post)
    elif title.lower().startswith('show hn'):
        show_posts.append(post)
    else:
        other_posts.append(post)

print('Total Ask HN Posts: ' + str(len(ask_posts)))
print('Total Show HN Posts: ' + str(len(show_posts)))
print('Total Other HN Posts: ' + str(len(other_posts)))

if len(ask_posts)+len(show_posts)+len(other_posts) == len(hn):
    print('Sum of lists = Length of hn  CHECK')

Total Ask HN Posts: 1744
Total Show HN Posts: 1162
Total Other HN Posts: 17194
Sum of lists = Length of hn  CHECK


In [3]:
# Calculate the average number of comments `Ask HN` posts receive.
total_ask_comments = 0

for post in ask_posts:
    total_ask_comments += int(post[4])
    
avg_ask_comments = total_ask_comments / len(ask_posts)
print('Average number of comments for "Ask HN" posts: ' + str(avg_ask_comments))

# Calculate the average number of comments `Show HN` posts receive.
total_show_comments = 0

for post in show_posts:
    total_show_comments += int(post[4])
    
avg_show_comments = total_show_comments / len(show_posts)
print('Average number of comments for "Show HN" posts: ' + str(avg_show_comments))

Average number of comments for "Ask HN" posts: 14.038417431192661
Average number of comments for "Show HN" posts: 10.31669535283993


On average, ask posts in our sample receive approximately 14 comments, whereas show posts receive approximately 10. Since ask posts are more likely to receive comments, we'll focus on these posts for the rest of the project.

# Finding the Amount of Ask Posts and Comments by Hour Created

In [4]:
import datetime as dt

result_list = [] # list of creation dates and number of comments per post in ask_posts

for post in ask_posts:
    created_at = post[6]
    num_comments = post[4]
    result_list.append([created_at, int(num_comments)])

counts_by_hour = {} # contains the number of ask posts created during each hour of the day.
comments_by_hour = {} # contains the corresponding number of comments ask posts created at each hour received.
date_format = "%m/%d/%Y %H:%M"

for row in result_list:
    date = row[0]
    comment = row[1]
    hour = dt.datetime.strptime(date, date_format).strftime("%H")
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = comment
    if hour in comments_by_hour:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += comment
comments_by_hour

{'00': 457,
 '01': 716,
 '02': 1384,
 '03': 422,
 '04': 340,
 '05': 493,
 '06': 398,
 '07': 269,
 '08': 497,
 '09': 257,
 '10': 794,
 '11': 643,
 '12': 691,
 '13': 1282,
 '14': 1419,
 '15': 4478,
 '16': 1831,
 '17': 1147,
 '18': 1441,
 '19': 1191,
 '20': 1724,
 '21': 1749,
 '22': 481,
 '23': 544}

# Calculating the Average Number of Comments for Ask HN Posts by Hour

In [5]:
avg_comments_per_hr = []

# Average comments per hour is calculted by dividing the
# total amount of comments by the amount of ask posts created per hour

for hour in comments_by_hour:
    avg_comments_per_hr.append([hour, comments_by_hour[hour]/counts_by_hour[hour]])

print("['Hour', Average # of Commments]")    
avg_comments_per_hr

['Hour', Average # of Commments]


[['11', 10.898305084745763],
 ['05', 10.48936170212766],
 ['09', 5.586956521739131],
 ['22', 6.680555555555555],
 ['17', 11.356435643564357],
 ['04', 7.083333333333333],
 ['07', 7.685714285714286],
 ['01', 11.737704918032787],
 ['13', 14.906976744186046],
 ['00', 8.160714285714286],
 ['12', 9.337837837837839],
 ['14', 13.13888888888889],
 ['23', 7.884057971014493],
 ['02', 23.45762711864407],
 ['16', 16.798165137614678],
 ['15', 38.27350427350427],
 ['03', 7.672727272727273],
 ['20', 21.28395061728395],
 ['08', 10.142857142857142],
 ['10', 13.233333333333333],
 ['19', 10.72972972972973],
 ['06', 8.844444444444445],
 ['18', 13.1],
 ['21', 15.9]]

# Sorting and Printing Values from a List of Lists

In [6]:
from operator import itemgetter

swap_avg_by_hr = []

for row in avg_comments_per_hr:
    swap_avg_by_hr.append([row[1],row[0]])

sorted_swap = sorted(swap_avg_by_hr, key=itemgetter(0), reverse=True)

sorted_swap

[[38.27350427350427, '15'],
 [23.45762711864407, '02'],
 [21.28395061728395, '20'],
 [16.798165137614678, '16'],
 [15.9, '21'],
 [14.906976744186046, '13'],
 [13.233333333333333, '10'],
 [13.13888888888889, '14'],
 [13.1, '18'],
 [11.737704918032787, '01'],
 [11.356435643564357, '17'],
 [10.898305084745763, '11'],
 [10.72972972972973, '19'],
 [10.48936170212766, '05'],
 [10.142857142857142, '08'],
 [9.337837837837839, '12'],
 [8.844444444444445, '06'],
 [8.160714285714286, '00'],
 [7.884057971014493, '23'],
 [7.685714285714286, '07'],
 [7.672727272727273, '03'],
 [7.083333333333333, '04'],
 [6.680555555555555, '22'],
 [5.586956521739131, '09']]

In [7]:
# Sort the values and print the the 5 hours with the highest average comments.

print("Top 5 Hours for 'Ask HN' Comments")
for avg, hr in sorted_swap[:5]:
    print(
        "{}: {:.2f} average comments per post".format(
            dt.datetime.strptime(hr, "%H").strftime("%H:%M"),avg
        )
    )

Top 5 Hours for 'Ask HN' Comments
15:00: 38.27 average comments per post
02:00: 23.46 average comments per post
20:00: 21.28 average comments per post
16:00: 16.80 average comments per post
21:00: 15.90 average comments per post


3:00pm is the hour that receives the most comments per post on average, with an average of 38.27 comments per post. There's about a 63% increase in the number of comments between the hours with the highest and second highest average number of comments.

# Conclusion

In this project, I analyzed ask posts and show posts to determine which type of post, and at what time, receives the most comments on average. Based on the analysis, to maximize the amount of comments a post receives, I recommend the post be categorized as an "Ask Post" and created between 3:00pm EST and 4:00pm EST.

Note: Posts with 0 comments were eliminated from the dataset.